core/src/main/java/feign/ReflectiveFeign.java

```diff
@@ -98,7 +98,16 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
       } else if ("toString".equals(method.getName())) {
         return toString();
       }
-      return dispatch.get(method).invoke(args);
+
+      try {
+          return dispatch.get(method).invoke(args);
+      } catch (Throwable th) {
+        Throwable cause = th.getCause();
+        if (th instanceof RetryableException && cause != null) {
+          throw cause;
+        }
+        throw th;
```

Comment:
This feels like we are breaking the previous signature for all feign methods and interfaces... before we would throw `FeignException`, now we throw `Exception`.

I think we could introduce a new method into the `Retryer` called `propagateException(FeignException)`, and ship feign with 2 implementations.  

* `Legacy` throws the current `FeignException`
* `Cause` throws the root cause for a `FeignException`

Reply:
This is a breaking change, we should weigh that carefully.

Should we consider a mechanism similar to Hystrix Javanica?

https://github.com/Netflix/Hystrix/blob/7f5a0afc23aa5ff82320560a04d4c81a45efd67c/hystrix-contrib/hystrix-javanica/src/main/java/com/netflix/hystrix/contrib/javanica/command/AbstractHystrixCommand.java#L148-L158

There is an assumption made that unless a particular option is set, the **original** exception is thrown.  If the original exception cannot be thrown, it is wrapped and thrown.  In Feign, we **always** wrap the exception.  I think this is the core issue @dinomite and I have been discussing in #737.  Maybe we can add a top level option or annotation to `Target` methods that allow for finer control over what exceptions get wrapped and what exceptions do not.

Reply:
@dinomite what are your thoughts on our suggestions?

Reply:
Thanks for reminding me that I had forgotten about this.

You're definitely right that this is big change, so doing it in a backwards-compatible way is the most tenable—I'd be ok with either @velo's suggestion of a pluggable solution or what you have suggested, @kdavisk6.  I'm a bit out of context on Feign recently, do either of these look better to you all?  Should  Does one of them sound preferable for me to give a go at?

Reply:
We will leave that up to you to choose.  The goal is to limit our exposure to existing clients. 

I think we can expand on @velo's suggestion to add a new method to the `Builder` called `propagateExceptions` which will change the default `Retryer` registered.  This will allow existing integrations to work and those that want the behavior to enable it without having to add additional code.  Thoughts?

Reply:
I liked it

Reply:
Got it.  I'll have some updates to this PR in the next week or two.